src/liballoc/vec.rs

```diff
@@ -2128,16 +2128,34 @@ impl<T> Vec<T> {
         //      for item in iterator {
         //          self.push(item);
         //      }
-        while let Some(element) = iterator.next() {
-            let len = self.len();
-            if len == self.capacity() {
-                let (lower, _) = iterator.size_hint();
-                self.reserve(lower.saturating_add(1));
-            }
-            unsafe {
-                ptr::write(self.get_unchecked_mut(len), element);
-                // NB can't overflow since we would have had to alloc the address space
-                self.set_len(len + 1);
+        let (lower, _) = iterator.size_hint();
+        self.reserve(lower);
+        loop {
+            let cap = self.capacity();
+            let result = iterator.by_ref().try_fold((), |(), element| {
```

Comment:
you can use `try_for_each` which takes a reference (so no `by_ref` needed) and doesn't require an element to fold through.

Reply:
Until https://github.com/rust-lang/rust/issues/46477 is fixed you probably should replace the closure by passing a free function instead (such as #62429 did for iterators). I think that's what @Mark-Simulacrum meant.

Reply:
> you can use try_for_each which takes a reference (so no by_ref needed) and doesn't require an element to fold through.

`try_for_each` just forwards to `try_fold` for `Chain` (same for most other iterators). I wanted to avoid getting the optimizer tripping up as much as possible as I have not seen any improvements so far.

`try_fold` also only takes a `&mut` so `by_ref` should definitely be removed as I just saw that `Iterator for &mut T` is unable to forward `try_fold` and friends due to object safety which I means I were just using the default implementation for `try_fold`, easily explaining why I am not seeing any improvement for chained iterators...

This might be fixable with some specialization at least https://github.com/rust-lang/rust/pull/68472

> Until #46477 is fixed you probably should replace the closure by passing a free function instead (such as #62429 did for iterators). I think that's what @Mark-Simulacrum meant.

Yeah, should do that. Was just trying to get an actual benchmark showing that this was an improvement first :/